packages/core/src/helpers/namespace.ts

```diff
@@ -1,3 +1,154 @@
+    paths: {
+        file: string;
+        origin: string;
+    };
+    packageInfo: PackageInfo;
+}
+
+export interface NamespaceBuilder {
+    (options: NamespaceBuilderParams): {
+        namespace: string;
+        hashPart: string;
+    };
+}
+
+export interface CreateNamespaceOptions {
+    prefix?: string;
+    hashSalt?: string;
+    hashFragment?: 'full' | 'minimal' | number;
+    buildNamespace?: NamespaceBuilder;
+    getPackageInfo?: (filePath: string) => PackageInfo;
+    handleNoMatch?: (
+        strict: boolean,
+        namespace: string,
+        filePath: string,
+        usedBy?: string
+    ) => string;
+    hashSeparator?: string;
+    strict?: boolean;
+    normalizePath: (dirPath: string, filePath: string) => string;
+    hashFn: (i: string) => string | number;
+}
+
+function defaultGetPackageInfo() {
+    return {
+        name: '',
+        version: '0.0.0',
+        dirPath: '',
+    };
+}
+
+export function defaultNamespaceBuilder({
+    prefix,
+    namespace,
+    hashSalt,
+    paths,
+    packageInfo,
+}: NamespaceBuilderParams) {
+    return {
+        namespace: prefix + namespace,
+        hashPart: hashSalt + packageInfo.name + '@' + packageInfo.version + '/' + paths.origin,
+    };
+}
+
+export function defaultNoMatchHandler(
+    strict: boolean,
+    ns: string,
+    stylesheetPath: string,
+    usedBy?: string
+): string {
+    throw new Error(
+        `Could not create namespace for:\n${stylesheetPath}\nthe last valid namespace tried was ${JSON.stringify(
+            ns
+        )}${usedBy ? ` that was used by:\n${usedBy}\n` : strict ? ' ' : ''}${
+            strict ? 'running on strict mode' : ''
+        }`
+    );
+}
+
+export function createNamespaceStrategy(options: CreateNamespaceOptions) {
+    const {
+        prefix = '',
+        hashSalt = '',
+        hashFragment = 'minimal',
+        buildNamespace = defaultNamespaceBuilder,
+        getPackageInfo = defaultGetPackageInfo,
+        handleNoMatch = defaultNoMatchHandler,
+        hashSeparator = '-',
+        strict = false,
+        hashFn,
+        normalizePath,
+    } = options;
+
+    const usedNamespaces = new Map<string, string>();
+
+    return (
+        namespace: string,
+        stylesheetOriginPath: string,
+        stylesheetPath: string = stylesheetOriginPath
+    ) => {
+        const packageInfo = getPackageInfo(stylesheetPath);
+        const buildNamespaceParams = {
+            prefix,
+            hashSalt,
+            namespace,
+            paths: {
+                file: packageInfo.dirPath
+                    ? normalizePath(packageInfo.dirPath, stylesheetPath)
+                    : stylesheetPath,
+                origin: packageInfo.dirPath
+                    ? normalizePath(packageInfo.dirPath, stylesheetOriginPath)
+                    : stylesheetOriginPath,
+            },
+            packageInfo,
+        };
+        const { namespace: resultNs, hashPart } = buildNamespace(buildNamespaceParams);
```

Comment:
`buildNamespace` might want to control just specific cases and will need to fallback to the default namespace. We talked about returning `undefined` to indicate that.